### PR TITLE
[Feature] implement cardinality(), element_at() and map_concat()

### PR DESCRIPTION
**Pull request by fzhedu**
## What type of PR is this:
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes:

Fixes #

## Problem Summary(Required):

implement cardinality(), element_at() and map_concat()

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [x] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 28 / 28 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AstToStringBuilder.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PolymorphicFunctionAnalyzer.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|12|12|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 29 / 29 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AstToStringBuilder.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PolymorphicFunctionAnalyzer.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|12|12|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java [hunk 1]
```diff
@@ -1019,6 +1019,15 @@ public Void visitFunctionCall(FunctionCallExpr node, Scope scope) {
                 argumentTypes = node.getChildren().stream().map(Expr::getType).toArray(Type[]::new);
             } else if (DecimalV3FunctionAnalyzer.argumentTypeContainDecimalV2(fnName, argumentTypes)) {
                 fn = DecimalV3FunctionAnalyzer.getDecimalV2Function(node, argumentTypes);
+            } else if (FunctionSet.CARDINALITY.equals(fnName)) {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java:1022]: you lose regesite `cardinality` signature? and don't need speical the implement in here

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java [hunk 2]
```diff
@@ -1161,7 +1170,17 @@ private void checkFunction(String fnName, FunctionCallExpr node, Type[] argument
                     }
                     break;
                 }
-
+                case FunctionSet.CARDINALITY: {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java:1173]: and don't need check

Comment (fzhedu) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java]: ok

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java (hunk 1 above)
Comment (fzhedu) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/ExpressionAnalyzer.java]: ok

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 17 / 17 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AstToStringBuilder.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PolymorphicFunctionAnalyzer.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 17 / 17 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AstToStringBuilder.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PolymorphicFunctionAnalyzer.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 17 / 17 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AstToStringBuilder.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PolymorphicFunctionAnalyzer.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|2|2|100.00%|[]|

**Review by Seaven — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Review by silverbullet233 — APPROVED** [review]

File: be/src/exprs/map_functions.cpp [hunk 3]
```diff
@@ -355,4 +355,133 @@ StatusOr<ColumnPtr> MapFunctions::distinct_map_keys(FunctionContext* context, co
...
+        not_null_columns.push_back(src_column);
+    }
+    if (not_null_columns.empty()) {
+        return ColumnHelper::create_const_null_column(chunk_size);
+    } else if (not_null_columns.size() == 1) {
+        return ColumnHelper::cast_to_nullable_column(not_null_columns[0]->clone_shared());
+    }
+
+    ssize_t columns_num = not_null_columns.size();
+    std::unique_ptr<uint32_t[]> hash_values[columns_num];
```
Comment (mofeiatwork) [inline be/src/exprs/map_functions.cpp:400]: why not use vector of vector

Comment (mofeiatwork) [reply be/src/exprs/map_functions.cpp:400]: is uint32_t enough to represent the hash ?

File: docs/sql-reference/sql-functions/array-functions/cardinality.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
+# cardinality
+
+## Description
+
+Returns the number of elements in the array. The result type is INT. if the parameter is NULL, the result is also NULL.
```
Comment (mofeiatwork) [inline docs/sql-reference/sql-functions/array-functions/cardinality.md:5]: what if null element

File: docs/sql-reference/sql-functions/array-functions/element_at.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
+# element_at
+
+## Description
+
+Returns the array element at the specific position. if any parameter is NULL, the result is also NULL.
+
+It is the alias of `[]` to get an element from an array.
```
Comment (mofeiatwork) [inline docs/sql-reference/sql-functions/array-functions/element_at.md:7]: why not use the `[]`

Comment (mofeiatwork) [reply docs/sql-reference/sql-functions/array-functions/element_at.md:7]: what's the use case of this `element_at` function

**Review by murphyatwork — APPROVED** [review]

File: be/src/exprs/map_functions.cpp (hunk 3 above)
Comment (fzhedu) [reply be/src/exprs/map_functions.cpp:400]: it is used to store the hash values from diffrerent keys, one array per map keys with various length. the inner array should get raw pointer to compute hash as L:409, learned in `array_distinct()`.

File: docs/sql-reference/sql-functions/array-functions/cardinality.md (hunk 4 above)
Comment (fzhedu) [reply docs/sql-reference/sql-functions/array-functions/cardinality.md:5]: it is a element, so counting it in size.

File: docs/sql-reference/sql-functions/array-functions/element_at.md (hunk 5 above)
Comment (fzhedu) [reply docs/sql-reference/sql-functions/array-functions/element_at.md:7]: trino.

**Comment by fzhedu** [comment]
@mergify backport branch-3.0

**Pull request merged**